Wait to raise Connected event for Read connection until container is caught up

markfields opened the pull request:
Part of #7884.  See comments in connectionStateHandler.ts for description of the desired behavior here.

Note that the new behavior is behind a feature gate "Fluid.Container.CatchUpBeforeDeclaringConnected".  I didn't make any changes to existing function `waitContainerToCatchUp`, even though some functionality is duplicated.  Once we finish rolling this out, we can remove the duplicate logic from that function since it will simply wait for the "connected" event (we may even deprecate it at that point)

To-Do:
- [ ] Need to update the README and specifically info about CatchingUp state

Some follow-up work items to open afterwards:
- [ ] Roll out new feature gate to enable new behavior for Loop partners
- [ ] Remove feature gate and deprecate `waitContainerToCatchUp`

markfields committed:
For read connection, wait until caught up before declaring "connected"

markfields committed:
Remove unnecessary read v. write check

vladsud commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -174,43 +197,22 @@ export async function waitContainerToCatchUp(container: IContainer) {
     return new Promise<boolean>((resolve, reject) => {
-        const deltaManager = container.deltaManager;
-
         container.on("closed", reject);
```
Looks like we never clear this callback on successful path - might be worth fixing it.

markfields replied:
Suggested @Abe27342 take care of this in PR https://github.com/microsoft/FluidFramework/pull/10071 since he's changing the callback.

markfields replied:
And he did - yay.

vladsud commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -174,43 +197,22 @@ export async function waitContainerToCatchUp(container: IContainer) {
+            waitForOps(container, resolve);
         };
         container.on(connectedEventName, callback);
 
```
can you please remove eslint comment below?

markfields replied:
Not going to touch it - `resume` has been deprecated and replaced with `connect`, @scottn12 will be replacing it soon (https://github.com/microsoft/FluidFramework/issues/9397)

vladsud commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -174,43 +197,22 @@ export async function waitContainerToCatchUp(container: IContainer) {
-            };
-            deltaManager.on("op", callbackOps);
-        };
+        //* I don't fully understand this comment
```
I think it points out to a fact that if we do not have checkpointSequenceNumber, then we can at least leverage join op sequence number for same purposes (for "write" connections).

What this code does not do a good job is dealing with disconnects while waiting for last op. I think it's worth opening a bug to incorporate that.

vladsud commented:
I think it looks good. Some things missing:
1. Tests
2. breaking.md note
3. Possibly: some feature gate to disable that change, in case we learn it screws up something we did not consider.

agarwal-navin commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -188,25 +189,29 @@ export class ConnectionStateHandler {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        const alreadyProcessedOwnJoinOp = this.handler.quorumClients()?.getMember(details.clientId) !== undefined;
```
I would call this `clientAlreadyInQuorum` and add a comment saying that if the client is in quorum, it would have processed it's own join op.

agarwal-navin commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -188,25 +189,29 @@ export class ConnectionStateHandler {
+            // and if we processed our Join op we would have processed our previous Leave op prior to that
+            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "Unexpected timer state" */);
+
+            // For write connections, we are caught up already (since we've seen our Join op) and this will be immediate
```
Based on the if condition above, this may not be true right. We can be in this else block if the connection is write and we haven't processed our join op.

agarwal-navin replied:
Ah okay, my bad. In the scenario I mentioned, we will go in the if condition. We will be here if it's a write connection and we have processed our join op.

agarwal-navin replied:
But this still may not be immediate. There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

agarwal-navin replied:
Also, please use `processed` our join op vs `seen` which mean different things.

markfields replied:
> There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

Yeah re-reading this change I came to the same conclusion, glad you saw it too, I was trying to figure out if I was missing something now or before when I wrote the code :P

agarwal-navin commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -188,25 +189,29 @@ export class ConnectionStateHandler {
+            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "Unexpected timer state" */);
+
+            // For write connections, we are caught up already (since we've seen our Join op) and this will be immediate
+            // For read connections, we will wait, so "Connected" event has identical semantics to write connection
```
I don't think "identical semantics to write connection" is entirely correct. It should explicitly mention what happens in read vs write. In both read and write, we wait to catch up to the sequence number we know about at this point in time. In write, we also wait for our own join op.
You should add this block to the big comment block before the if condition. And then maybe have comments in the if / else block specific to what happens in that scenario. Right now, its a little confusing.

markfields replied:
I made a bunch of changes to the comments, not sure I exactly addressed this but take a look and see if it needs further clarification?

agarwal-navin commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -157,6 +157,29 @@ export enum ConnectionState {
 }
 
+/** Invoke the callback once deltaManager has processed all ops known at the time this function is called */
+function waitForOps(container: IContainer, resolve: (hasCheckpointSequenceNumber: boolean) => void) {
```
`waitForOps` is too generic. Maybe this should be called `waitToProcessKnownOps`?

agarwal-navin replied:
Or `catchUpToKnownOps`

agarwal-navin commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -18,6 +18,7 @@ export interface IConnectionStateHandler {
     maxClientLeaveWaitTime: number | undefined,
     logConnectionIssue: (eventName: string) => void,
     connectionStateChanged: () => void,
+    onCaughtUp: (callback: () => {}) => void,
```
nit: `onCaughUpToKnownOps`.

markfields commented:
Revisiting this finally.  Writing down some questions I'm trying to clarify in my head:

* How does `startJoinOpTimer` work?  When it fires, it doesn't do anything.  I guess `receivedAddMemberEvent` keys off whether it's set or not, and if so may call `applyForConnectedState` which in turn may call `this.setConnectionState(ConnectionState.Connected);`.
* For write connections, does `waitContainerToCatchUp` actually cover the scenario of leveraging join op to indicate "caught up"?  The comment in there (the one I said I don't understand) seems to indicate so, but I don't see it.

markfields committed:
PR Feedback so far

markfields commented:
@jatgarg can you take a look too please? Thanks!

markfields commented:
Answering my own question...

> For write connections, does waitContainerToCatchUp actually cover the scenario of leveraging join op to indicate "caught up"? The comment in there (the one I said I don't understand) seems to indicate so, but I don't see it.

`waitContainerToCatchUp` gates on `connected` event which for write connections means we've processed own join op.

This PR kind of inverts this - now _always_ "connected" event means we're caught up.  So `waitContainerToCatchUp` is actually "wait for the container to be connected".

markfields committed:
Simplify waitContainerToCatchUp.

It can just gate on "connected" state,
since "connected" state now gates on caught up.

markfields committed:
Fix bug where we didn't always wait for catch up before "connected" state

markfields committed:
PR feedback

markfields commented:
@vladsud, @agarwal-navin, @jatgarg -- Here's my update on this PR, and some input I need from you:

* Do we all agree with the specification that "connected" means "caught up to known ops" even for write clients? i.e. that's an additional delay beyond the "join" op now.  Also, the way I've implemented so far the definition of "known ops" varies based on when a particular callback is invoked, so it's a little wishy-washy.  Thoughts?
* I started looking into a feature gate for this.  Ideally the gate would be used in two places: (1) ConnectionStateHandler to determine when to transition to "connected" state and (2) waitContainerToCatchUp can just wait for "connected" event when the gate is enabled (don't need to wait for additional ops).  I don't see a way to cover (2) because it's a function taking in an `IContainer`, rather than being a private function on `Container` class.  I think this will be fine for `waitContainerToCatchUp` to essentially catch up twice - if `hasCheckpointSequenceNumber` it'll be a no-op the second time anyway.
  * If we agree then I'll revert behavior of `waitContainerToCatchUp` back to how it was.
* I need to write tests
* I need to move this to `next` and include BREAKING.md note.

Anything else I should consider here?  I anticipate moving this from Draft on Monday or Tuesday.  Thanks guys!

vladsud commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -127,6 +128,10 @@ export class ConnectionStateHandler {
     }
 
+    private transitionToConnectedStateWhenCaughtUp() {
+        this.handler.onCaughUpToKnownOps(() => this.setConnectionState(ConnectionState.Connected));
```
I do not think this is correct flow. In a sense that it would work only if we do not lose connection while processing ops.
If we do, we have to abandon this process and start over.

jatgarg commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -613,6 +613,9 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
                         this.propagateConnectionState();
                     }
                 },
+                onCaughUpToKnownOps: (callback: () => void) => {
```
nit: onCaughUpToKnownOps -> onCaughtUpToKnownOps

jatgarg commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -613,6 +613,9 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
                         this.propagateConnectionState();
                     }
                 },
+                onCaughUpToKnownOps: (callback: () => void) => {
```
The signature you added above for "callback" is different.

markfields replied:
Yeah this code doesn't care about the boolean parameter in the callback so just omit it.  The function types are compatible (the argument will just be dropped by JS).

jatgarg commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -174,43 +198,19 @@ export async function waitContainerToCatchUp(container: IContainer) {
-            };
-            deltaManager.on("op", callbackOps);
-        };
+        const usedCheckpointSequenceNumber = container.deltaManager.hasCheckpointSequenceNumber;
```
"usedCheckpointSequenceNumber " does not seem right, I think it should just be "hasCheckpointSequenceNumber"

markfields committed:
typo

markfields committed:
Some comments / renames

markfields committed:
More clarity (ish)

markfields committed:
Note to self

markfields committed:
Introduce CatchUpWaiter class

markfields committed:
Switch to callback instead of event for CatchUpWaiter

markfields committed:
Reshuffling stuff

markfields committed:
Finalize shape and name of CatchUpMonitor

markfields committed:
ICatchupMonitor interface and Immediate impl for when flight is disabled

markfields commented:
@agarwal-navin @jatgarg ready for Review - Except I will rebase onto next, and I will write a bunch of tests.  Thanks!

tyler-cai-microsoft commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,91 @@
+    (event: "caughtUp", listener: CaughtUpListener): void;
+}
+
+export interface ICatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents>, IDisposable { }
```
What does the catchUpMonitor monitor catching up to?

markfields replied:
I'll add a comment, thanks

tyler-cai-microsoft commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -9,27 +9,53 @@ import { ConnectionMode, IQuorumClients, ISequencedClient } from "@fluidframewor
 }
 
-const JoinOpTimer = 45000;
+const JoinOpTimeout = 45000;
```
`JoinOpTimeoutMs`

markfields committed:
Merge branch 'main' into connected-event-for-read

markfields committed:
lock

markfields committed:
Adding some comments, some cleanup

tyler-cai-microsoft commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -164,49 +193,87 @@ export class ConnectionStateHandler {
+    }
+
+    private setConnectionState(value: ConnectionState.Disconnected, reason: string): void;
+    private setConnectionState(value: ConnectionState.Connected): void;
```
I'm assuming these are do nothing calls?

tyler-cai-microsoft commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -164,49 +193,87 @@ export class ConnectionStateHandler {
+        } else {
+            // We're not waiting for Leave or Join op, but we do need to wait until we are caught up (to now-known ops)
+            // before transitioning to Connected state.
+            this.catchUpMonitor.on("caughtUp", this.transitionToConnectedState);
```
Is it worth turning this magic string into an exported variable?

markfields replied:
No need, It's protected by the type system 😎

tyler-cai-microsoft commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,107 @@
+            if (event === "caughtUp") {
+                const caughtUpListener = listener as CaughtUpListener;
+                if (this.caughtUp) {
+                    caughtUpListener(this.hasCheckpointSequenceNumber);
```
where does `caughtUpListener` come from?

markfields replied:
It's the callback provided by whoever is calling `monitor.on("caughtUp")`  This is a trick so we can hide the `caughtUp` state - the other approach is anyone who ever registers for "caughtUp" needs to first check to see if we already are.

markfields committed:
Merge remote-tracking branch 'origin/main' into connected-event-for-read

markfields committed:
lock files after merge main and install

markfields committed:
Build fixes post-merge

markfields committed:
Minor updates, and write notes on what tests to implement

markfields committed:
lock update on install

markfields committed:
Trying to add some tests.

I updated CatchUpMonitor to take a DM instead of Container.
This is not good because DM has to be connected
but you can't ask it if it is.

New plan: Revert all changes to waitContainerToCatchUp.
Move CatchUpMonitor into ConnectionStateHandler,
and have it take from Container
some subset of DeltaManager
which can be replaced with "immediate"
implementation when feature gate is off.

This approach is best because the lifecycle
of the CatchUpMonitor is tightly coupled
to Container's connection state/lifecycle.

markfields committed:
revert changes to waitContainerToCatchUp.

Add comment explaining the new semantics of Connected state.

markfields committed:
Pass in deltaManager to receivedConnectEvent

This way CatchUpMonitor can be totally contained within
connectionStateHandler.

markfields committed:
Switch to optional deltaManager parameter

markfields committed:
0x2a6 test

markfields committed:
Some cleanup/clarifying refactors

markfields committed:
Write broken test

markfields committed:
Fix case where assert 0x2a6 fires for write connection

markfields committed:
Fix typo

markfields committed:
Add a test for when quorum is not yet initialized when "connect" happens

markfields committed:
Logging changes

markfields committed:
Lint fix and rename

markfields committed:
Merge branch 'main' into fix-assert-0x2a6

markfields committed:
lock file changes after a clean install

markfields committed:
Build fixes post-merge

markfields committed:
Merge branch 'main' into fix-assert-0x2a6

markfields committed:
Merge branch 'main' into fix-assert-0x2a6

markfields committed:
PR feedback

markfields committed:
Revert unrelated lock file changes

markfields committed:
nit

markfields committed:
Merge remote-tracking branch 'origin/main' into fix-assert-0x2a6

markfields committed:
Merge branch 'fix-assert-0x2a6' into connected-event-for-read

markfields committed:
Merge remote-tracking branch 'origin/main' into connected-event-for-read

markfields committed:
Minor cleanup post-merge

markfields committed:
FIx build in tests

markfields committed:
Unit Tests are green

(but one is still skipped and I may write a few more)

markfields commented:
Remaining tasks:
- [ ] Write UTs for CatchUpMonitor and ImmediateCatchUpMonitor
- [ ] Fix the skipped test in Container.spec.ts
- [ ] Address //* comments

markfields committed:
Minor clean up / fixes

markfields commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -178,6 +178,10 @@ export async function waitContainerToCatchUp(container: IContainer) {
         container.on("closed", closedCallback);
 
+        // Depending on config, transition to "connected" state may include the guarantee
+        // that all known ops have been processed.  If so, we may introduce additional wait here.
```
What do you all think - is this correct or not?  I may not add additional delay since events fire synchronously.

markfields commented:
@agarwal-navin @vladsud @tyler-cai-microsoft @jatgarg -- Thanks for the early review!  Ready for review now.  Just need to add UTs for the new "CatchUpMonitor" classes but otherwise ready to go.

jatgarg commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1508,9 +1512,15 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             }
 
+            const deltaManagerForCatchingUp =
+                this.mc.config.getBoolean("Fluid.Container.CatchUpBeforeDeclaringConnected") === true ?
```
Have you given a thought that it should be opt out and not opt in?

markfields replied:
Only a little.  I think shipping dark is the standard operating procedure, and there's no rush to diverge from this.  Also I personally want to practice what it's like to use feature gates.

markfields committed:
Revert unrelated changes

markfields committed:
CatchUpMonitor tests

markfields committed:
Merge branch 'main' into connected-event-for-read

markfields commented on packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:
```diff
@@ -91,8 +106,25 @@ describe("ConnectionStateHandler Tests", () => {
+        deltaManagerForCatchingUp = new MockDeltaManagerForCatchingUp();
     });
 
+    /** Test plan:
```
Note to self - delete this block

jatgarg commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1509,9 +1513,15 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             }
 
+            const deltaManagerForCatchingUp =
+                this.mc.config.getBoolean("Fluid.Container.CatchUpBeforeDeclaringConnected") === true ?
```
nit: No need to check "=== true"

markfields replied:
I prefer to be explicit that it must be set to true (but you're right the code works either way)

jatgarg commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -210,32 +217,49 @@ export class ConnectionStateHandler {
+
     /**
-     * The "connect" event indicates the connection to the Relay Service is live.
+     * The "connect" event indicates the connection to the ordering service is live.
```
We are moving towards "relay service" terminology than "ordering service". So, we should revert this change.

jatgarg commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -210,32 +217,49 @@ export class ConnectionStateHandler {
      * @param connectionMode - Read or Write connection
      * @param details - Connection details returned from the ordering service
+     * @param deltaManager - DeltaManager to be used for delaying Connected transition until caught up.
+     * If it's undefined, then don't delay and transition to Connected event as soon as Leave/Join op are accounted for
```
nit: "Connected event" -> "Connected state"

jatgarg commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -210,32 +217,49 @@ export class ConnectionStateHandler {
+            inConnectingState,
+            this.logger,
+            { eventName: "cannotTransitionToConnectedState", connectionState: ConnectionState[this._connectionState] });
+        if (!inConnectingState) {
```
Shouldn't this be an invariant and should we throw on it?

jatgarg commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -248,11 +272,16 @@ export class ConnectionStateHandler {
 
         // For write connections, this pending clientId could be in the quorum already (i.e. join op already processed).
-        // We are fetching ops from storage in parallel to connecting to Relay Service,
+        // We are fetching ops from storage in parallel to connecting to ordering service,
```
Same as above.

jatgarg commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+export class ImmediateCatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents> implements ICatchUpMonitor {
+    constructor() {
+        super();
+        this.on("newListener", (event: string, listener) => {
```
Cann't we just add the type of listener here? instead of typecasting later?

markfields replied:
I like it as-is because if there were other events supported here the listener would be one of many types (probably a union type) and then you'd need to narrow it down depending on the event.

jatgarg commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+        this.deltaManager.on("op", this.opHandler);
+
+        // Simulate the last processed op to set caughtUp in case we already are
+        this.opHandler({ sequenceNumber: this.deltaManager.lastSequenceNumber });
```
I think we should change the order of statements here on line 52 and 55. So like initialize first before registering the handler.

jatgarg commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+            if (event === "caughtUp") {
+                const caughtUpListener = listener as CaughtUpListener;
+                if (this.caughtUp) {
+                    caughtUpListener(this.hasCheckpointSequenceNumber);
```
You are passing "transitionToConnectedState" as "caughtUpListener", but transitionToConnectedState does not take any argument. So, what is the logic here of passing the "hasCheckpointSequenceNumber".

markfields replied:
I went back and forth on this.  In the future when this feature gate is at 100%, `waitContainerToConnect` will become obsolete, replaced by simply waiting for the `connected` event.  If someone were depending on the boolean return value of that function, maybe we would want to update `connected` to include that in its event args.

I feel it's kind of a strange thing to expose, I guess the point is some kind of "confidence" indicator of how caught up you'll be when the thing resolves/fires.  @vladsud thoughts on maintaining this info long-term (and whether it belongs in the short term)

vladsud replied:
I'm fine not exposing this info anywhere. It used to be that FRS was an outlier here as it did not have checkpoint sequence number. But with recent work to expose sequence numbers on join & leave signals, we should ensure that both FRS is on par with ODSP, as this feature is just too good not to have it - it allows many things beyond ability to catch up, like building sync primitives (including leader / summarizer election) using signals only. Plus, it was added to solve race condition on connection that resulted in client missing ops - so it's really not that optional to have it.

vladsud commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+        this.opHandler({ sequenceNumber: this.deltaManager.lastSequenceNumber });
+
+        // If a listener is added after we are already caught up, notify that new listener immediately
+        this.on("newListener", (event: string, listener) => {
```
It feels like instead of using events, it's better to use a property that exposes a promise. I.e. something like
```ts
public caughtUp: Promise<number>;
```
Promise seems to be better because this event files once, and past "firing" reflected on new listeners - that's exactly what promise does.

markfields replied:
That is a great point.  I don't remember why I switched to event-based to begin with, but I bet it would work better with Promise.  That said, I don't have time to try it out now, and also leaving it will reduce merge conflicts with your follow-on refactor.

markfields commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+export class ImmediateCatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents> implements ICatchUpMonitor {
+    constructor() {
+        super();
+        this.on("newListener", (event: string, listener) => {
```
This is just a notification after the listener is added. It'll get removed when `dispose` is called below which happens on disconnect.

markfields committed:
Merge remote-tracking branch 'origin/main' into connected-event-for-read

markfields commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -184,7 +188,10 @@ export class ConnectionStateHandler {
             && !this.waitingForLeaveOp
         ) {
             this.waitEvent?.end({ source });
-            this.setConnectionState(ConnectionState.Connected);
```
For the record - Vlad prototyped this in #10958 and is thinking about doing the refactor after this PR is merged.

markfields commented on packages/loader/container-loader/src/connectionStateHandler.ts:
```diff
@@ -210,32 +217,49 @@ export class ConnectionStateHandler {
         this.setConnectionState(ConnectionState.Disconnected, reason);
     }
 
+    private readonly transitionToConnectedState = () => {
```
Prototype of this direction is in #10958 - thanks for pursuing this, definitely an improvement!

markfields committed:
PR Feedback

vladsud reviewed (approved):
Singing off to unblock you - I did not look through it now, but was poking in my prototyping on top of it and overall, it looks fine.

markfields commented on packages/loader/container-loader/src/catchUpMonitor.ts:
```diff
@@ -0,0 +1,101 @@
+    ) {
+        super();
+
+        this.hasCheckpointSequenceNumber = this.deltaManager.hasCheckpointSequenceNumber;
```
getting rid of hasCheckpointSequenceNumber makes it a little more generic, I'll see about adding some comments or any other asserts etc

markfields replied:
I looked at adding `deltaManager.on("disconnect", () => this.dispose())`, which I do like.  I am sorry I am just out of time though.  I would want to add tests, etc.  It's an easy change to tack on later.

Also, I think it works across connections anyway.

markfields committed:
Fix test

markfields commented:
Opened https://dev.azure.com/fluidframework/internal/_workitems/edit/1384 to track rolling this out
